netaddr.go

```diff
@@ -412,96 +332,118 @@ func (ip IP) Is6() bool {
 // returns the wrapped IPv4 address. Otherwise it returns ip, regardless
 // of its type.
 func (ip IP) Unmap() IP {
-	if !ip.Is4in6() {
-		return ip
+	if ip.Is4in6() {
+		ip.z = z4
 	}
-	a := ip.ipImpl.as16()
-	return IP{v4Addr{a[12], a[13], a[14], a[15]}}
+	return ip
 }
 
+// zmu guards uniqZone, a weakref map of *zones by zoneName.
+var (
+	zmu      sync.Mutex
+	uniqZone = map[string]uintptr{} // zone name to its uintptr(*zone)
+)
+
 // WithZone returns an IP that's the same as ip but with the provided
-// zone. If zone is empty, the zone is removed. If ip is an IPv4
+// zone. If zoneName is empty, the zone is removed. If ip is an IPv4
 // address it's returned unchanged.
-func (ip IP) WithZone(zone string) IP {
-	if zone == "" {
-		if z, ok := ip.ipImpl.(v6AddrZone); ok {
-			return IP{z.v6Addr}
-		}
+func (ip IP) WithZone(zoneName string) IP {
+	if !ip.Is6() {
 		return ip
 	}
-	switch ip := ip.ipImpl.(type) {
-	case v6Addr:
-		return IP{v6AddrZone{ip, zone}}
-	case v6AddrZone:
-		return IP{v6AddrZone{ip.v6Addr, zone}}
+	if zoneName == "" {
+		ip.z = z6noz
+		return ip
 	}
+
+	zmu.Lock()
+	defer zmu.Unlock()
+
+	addr, ok := uniqZone[zoneName]
+	var z *zone
+	if ok {
+		z = (*zone)((unsafe.Pointer)(addr))
+	} else {
+		z = &zone{name: zoneName}
+		uniqZone[zoneName] = uintptr(unsafe.Pointer(z))
```

Comment:
As this is outside the Go specification, I think:

1. It deserves commentary explaining that fact somewhere.
2. It would be good to add a check that could catch a potentially novel runtime and panic. A cheap one would be to put the zone uintptr in the *zone as an object header:

```
type zone struct {
    self uintptr
    ...
}
```

Then when using a zone, first check that `uintptr(unsafe.Pointer(z))==z.self` and panic if it doesn't.